java/ql/src/experimental/Security/CWE/CWE-730/RegexInjection.ql

```diff
@@ -0,0 +1,101 @@
+/**
+ * @name Regular expression injection
+ * @description User input should not be used in regular expressions without first being sanitized,
+ *              otherwise a malicious user may be able to provide a regex that could require
+ *              exponential time on certain inputs.
+ * @kind path-problem
+ * @problem.severity error
+ * @precision high
+ * @id java/regex-injection
+ * @tags security
+ *       external/cwe/cwe-730
+ *       external/cwe/cwe-400
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/**
+ * A data flow sink for untrusted user input used to construct regular expressions.
+ */
+class RegexSink extends DataFlow::ExprNode {
+  RegexSink() {
+    exists(MethodAccess ma, Method m | m = ma.getMethod() |
+      (
+        m.getDeclaringType().hasQualifiedName("java.lang", "String") and
+        (
+          ma.getArgument(0) = this.asExpr() and
+          (
+            m.hasName("matches") or
+            m.hasName("split") or
+            m.hasName("replaceFirst") or
+            m.hasName("replaceAll")
+          )
+        )
+        or
+        m.getDeclaringType().hasQualifiedName("java.util.regex", "Pattern") and
+        (
+          ma.getArgument(0) = this.asExpr() and
+          (
+            m.hasName("compile") or
+            m.hasName("matches")
+          )
+        )
+        or
+        m.getDeclaringType().hasQualifiedName("org.apache.commons.lang3", "RegExUtils") and
+        (
+          ma.getArgument(1) = this.asExpr() and
+          m.getParameterType(1).(Class).hasQualifiedName("java.lang", "String") and
```

Comment:
You could use the existing [`TypeString`](https://codeql.github.com/codeql-standard-libraries/java/semmle/code/java/JDK.qll/type.JDK$TypeString.html) here:
```suggestion
          m.getParameterType(1) instanceof TypeString and
```